### DISCUSSION TITLE
Allow Specifying Additional PodMonitor Labels

### DISCUSSION THREAD

**Discussion by aphistic** [cloudnative-pg/cloudnative-pg#1617] [Ideas]
Hi!

I'd like the ability to provide additional labels that are added to a `Cluster`'s `PodMonitor` when it's created.

In my cluster, I'm using the Prometheus Operator to set up scraping, and the `Prometheus` instance is set up with specific `PodMonitor` or `ServiceMonitor` labels to specify which Prometheus instance should scrape it.

I had some thoughts about something like:
```
monitoring:
  podMonitor:
    enabled: true
    additionalLabels:
      someLabel: withValue
```

But that would duplicate the existing `enablePodMonitor` value, so I thought I'd create this and see what you all think!

**Comment by itay-grudev** [comment]
I have a similar situation. In my case, I need to specify the kube-prometheus-stack  release name in the `PodMonitor` labels or it won't get scraped. My case requires the following:

```yaml
monitoring:
  podMonitor:
    enabled: true
    labels:
      release: prometheus
```

P.S. I suggest calling it just `labels` as opposed to `additionalLabels` as that's what most charts use.

**Comment by sxd** [comment]
@aphistic @itay-grudev hi!

What about following the same behavior like the one described [here](https://cloudnative-pg.io/documentation/1.19/labels_annotations/#defining-clusters-metadata) or you think that the PodMonitor will require their own specific labels?

**Comment by aphistic** [comment]
@sxd Oh! Are those labels also applied to `PodMonitor` (or any resource related to the cluster)? If that's the case, I don't think there needs to be a separate set of options, but it might make things cleaner by avoiding superfluous labels.

In my case, the labels I apply are something like `metrics.mydomain.com/scrapeScope=cluster` or `metrics.mydomain.com/scrapeScope=postgres` if I want a different Prometheus instance ("cluster" vs. "postgres") to scrape the metrics. These labels are set, for example, via the `podMonitorSelector` in the Prometheus Operator's [`Prometheus` CRD](https://prometheus-operator.dev/docs/operator/api/#monitoring.coreos.com/v1.Prometheus).

Those labels could be added to the cluster, but then you might have `metrics.mydomain.com/scrapeScope=postgres` on the `StatefulSet` where it wouldn't be needed/might add confusion.

**Comment by sxd** [comment]
@aphistic well, yes the labels are inherited by the podmonitor object too https://github.com/cloudnative-pg/cloudnative-pg/blob/main/controllers/cluster_create.go#L817 can you try this feature and check if that works for you ?

We don't use a StatefulSet so, no it will not hit any StatefulSet =)

Please let us know, in the meantime I'm turning this into a discussion!

Regards,

**Reply by itay-grudev** [reply]
I don't think the `Cluster` labels were applied to my `PodMonitor`. I also think it would be better to have separate labels for the `PodMonitor` resource, as they most likely don't apply to the rest of the resources.

**Comment by Magicloud** [comment]
I do not think that "labels" are inherited. With manifest:
```
apiVersion: postgresql.cnpg.io/v1
kind: Cluster
metadata:
  name: dify
  labels:
    release: monitoring
spec:
  instances: 1
  storage:
    size: 10Gi
  monitoring:
    enablePodMonitor: true
```
I got (Non-related lines omitted):
```
$ kubectl -ndify get pod dify-1 -oyaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    cnpg.io/nodeSerial: "1"
    cnpg.io/operatorVersion: 1.25.1
  labels:
    cnpg.io/cluster: dify
    cnpg.io/instanceName: dify-1
    cnpg.io/instanceRole: primary
    cnpg.io/podRole: instance
    role: primary
  name: dify-1
  namespace: dify
```
And
```
$ kubectl -ndify get podmonitor dify -oyaml
apiVersion: monitoring.coreos.com/v1
kind: PodMonitor
metadata:
  annotations:
    cnpg.io/operatorVersion: 1.25.1
  creationTimestamp: "2025-03-18T12:08:01Z"
  generation: 1
  labels:
    cnpg.io/cluster: dify
  name: dify
  namespace: dify
  ownerReferences:
  - apiVersion: postgresql.cnpg.io/v1
    controller: true
    kind: Cluster
    name: dify
    uid: 14a21d25-bea5-403c-aa32-78613f2f1aaf
  resourceVersion: "4852518"
  uid: 4eb70204-5647-46ed-b623-77c080bb7b37
spec:
  namespaceSelector: {}
  podMetricsEndpoints:
  - bearerTokenSecret:
      key: ""
      name: ""
    port: metrics
  selector:
    matchLabels:
      cnpg.io/cluster: dify
```

I am using 1.25.1.
